packages/interface-stream-muxer/src/stream.ts

```diff
@@ -0,0 +1,321 @@
+    }
+
+    this.onSinkEnd()
+  }
+
+  // Close for reading and writing (local error)
+  abort (err: Error): void {
+    log.trace('%s stream %s abort', this.stat.direction, this.id, err)
+    // End the source with the passed error
+    this.streamSource.end(err)
+    this.abortController.abort()
+    this.onSinkEnd(err)
+  }
+
+  // Close immediately for reading and writing (remote error)
+  reset (): void {
+    const err = new CodeError('stream reset', ERR_STREAM_RESET)
+    this.resetController.abort()
+    this.streamSource.end(err)
+    this.onSinkEnd(err)
+  }
+
+  async sink (source: Source<Uint8ArrayList | Uint8Array>): Promise<void> {
+    if (this.sinkSunk) {
+      throw new CodeError('sink already called on stream', ERR_DOUBLE_SINK)
+    }
+
+    this.sinkSunk = true
+
+    if (this.sinkEnded) {
+      throw new CodeError('stream closed for writing', ERR_SINK_ENDED)
+    }
+
+    const signal = anySignal([
+      this.abortController.signal,
+      this.resetController.signal,
+      this.closeController.signal
+    ])
+
+    try {
+      source = abortableSource(source, signal)
+
+      if (this.stat.direction === 'outbound') { // If initiator, open a new stream
+        this.sendNewStream()
+      }
+
+      for await (let data of source) {
+        while (data.length > 0) {
+          if (data.length <= this.maxDataSize) {
+            this.sendData(data instanceof Uint8Array ? new Uint8ArrayList(data) : data)
+            break
+          }
+          data = data instanceof Uint8Array ? new Uint8ArrayList(data) : data
+          this.sendData(data.sublist(0, this.maxDataSize))
+          data.consume(this.maxDataSize)
+        }
+      }
+    } catch (err: any) {
+      if (err.type === 'aborted' && err.message === 'The operation was aborted') {
+        if (this.closeController.signal.aborted) {
+          return
+        }
+
+        if (this.resetController.signal.aborted) {
+          err.message = 'stream reset'
+          err.code = ERR_STREAM_RESET
+        }
+
+        if (this.abortController.signal.aborted) {
+          err.message = 'stream aborted'
+          err.code = ERR_STREAM_ABORT
+        }
+      }
+
+      // Send no more data if this stream was remotely reset
+      if (err.code === ERR_STREAM_RESET) {
+        log.trace('%s stream %s reset', this.stat.direction, this.id)
+      } else {
+        log.trace('%s stream %s error', this.stat.direction, this.id, err)
+        try {
+          this.sendReset()
+          this.stat.timeline.reset = Date.now()
+        } catch (err) {
+          log.trace('%s stream %s error sending reset', this.stat.direction, this.id, err)
+        }
+      }
+
+      this.streamSource.end(err)
+      this.onSinkEnd(err)
+      return
+    } finally {
+      signal.clear()
+    }
+
+    try {
+      this.sendCloseWrite()
+    } catch (err) {
+      log.trace('%s stream %s error sending close', this.stat.direction, this.id, err)
```

Comment:
Do you think we should include the `direction` in these trace logs?

Reply:
It's in there - `this.stat.direction`